### PR TITLE
[SPARK-49683][SQL] Block trim collation

### PR DESCRIPTION
**Pull request by jovanpavl-db**
### What changes were proposed in this pull request?
Trim collation is currently in implementation phase. These change blocks all paths from using it and afterwards trim collation gets enabled for different expressions it will be gradually whitelisted.

### Why are the changes needed?
Trim collation is currently in implementation phase. These change blocks all paths from using it and afterwards trim collation gets enabled for different expressions it will be gradually whitelisted.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
No additional tests, just added field that's not used.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala [hunk 1]
```diff
@@ -23,41 +23,73 @@ import org.apache.spark.sql.types.{AbstractDataType, DataType, StringType}
...
-    other.isInstanceOf[StringType] && other.asInstanceOf[StringType].isNonCSAI
+    other.isInstanceOf[StringType] && other.asInstanceOf[StringType].isNonCSAI &&
+      canUseTrimCollation(other)
 }
+
+object StringTypeNonCSAICollation extends StringTypeNonCSAICollation(false) {
+  def apply(supportsTrimCollation: Boolean): StringTypeNonCSAICollation = {
+    new StringTypeNonCSAICollation(supportsTrimCollation)
+  }
+}
```
Comment (HyukjinKwon) [inline sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala:95]:
```suggestion
}

```

**Review by HyukjinKwon — APPROVED** [review]

**Review by MaxGekk — COMMENTED** [review]
@uros-db @mihailom-db Could you review this PR, please.

File: sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala [hunk 2]
```diff
@@ -23,41 +23,76 @@ import org.apache.spark.sql.types.{AbstractDataType, DataType, StringType}
...
   override private[sql] def acceptsType(other: DataType): Boolean =
-    other.isInstanceOf[StringType] && other.asInstanceOf[StringType].supportsBinaryEquality
+    other.isInstanceOf[StringType] && other.asInstanceOf[StringType].supportsBinaryEquality &&
+      canUseTrimCollation(other)
+}
+
+object StringTypeBinary extends StringTypeBinary(false) {
+  def apply(supportsTrimCollation: Boolean): StringTypeBinary = {
+    new StringTypeBinary(supportsTrimCollation)
+  }
```
Comment (mihailom-db) [inline sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala:47]: I am not sure you need to change this from case object to case class. Can you try to for case object to extend from AbstractStringType(false)

Comment (uros-db) [reply sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala:47]: I would just consider adding an apply method with bool param

File: sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala [hunk 3]
```diff
@@ -23,41 +23,76 @@ import org.apache.spark.sql.types.{AbstractDataType, DataType, StringType}
 /**
  * AbstractStringType is an abstract class for StringType with collation support.
```
Comment (uros-db) [inline sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala:24]: let's update these comments to reflect the new state of things - adding trimming support

**Review by uros-db — APPROVED** [review]
lgtm

**Review by mihailomilosevic2001 — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**